linfa-ensemble/src/random_forest/algorithm.rs

```diff
@@ -0,0 +1,191 @@
+        let ntrees = self.hyperparameters.n_estimators;
+        assert!(ntrees > 0, "Run .fit() method first");
+
+        let mut result: Vec<u64> = Vec::with_capacity(x.nrows());
+        let flattened: Vec<Vec<u64>> = self
+            .trees
+            .iter()
+            .map(|tree| tree.predict(&x).unwrap().to_vec())
+            .collect();
+
+        for sample_idx in 0..x.nrows() {
+            // hashmap to store most common prediction across trees
+            let mut counter_stats: HashMap<u64, u64> = HashMap::new();
+            for sp in &flattened {
+                *counter_stats.entry(sp[sample_idx]).or_insert(0) += 1;
+            }
+
+            // aggregate counters to final prediction
+            let final_pred = counter_stats
+                .iter()
+                .max_by(|a, b| a.1.cmp(&b.1))
+                .map(|(k, _v)| k)
+                .unwrap();
+
+            result.push(*final_pred);
+        }
+        Ok(Array1::from(result))
+    }
+}
+
+impl ProbabilisticPredictor for RandomForest {
+    /// Return probability of predicted class for each sample, calculated as the rate of independent trees that
+    /// have agreed on such prediction
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - A 2D array of floating point elements
+    ///
+    ///
+    fn predict_probabilities(
+        &self,
+        x: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+    ) -> Result<Vec<Array1<f64>>, LinfaError> {
+        let ntrees = self.hyperparameters.n_estimators;
+        assert!(ntrees > 0, "Run .fit() method first");
+
+        let nclasses = self.trees[0].hyperparameters().n_classes;
+        let mut result: Vec<Array1<f64>> = Vec::with_capacity(x.nrows());
+
+        let flattened: Vec<Vec<u64>> = self
+            .trees
+            .iter()
+            .map(|tree| tree.predict(&x).unwrap().to_vec())
+            .collect();
+
+        for sample_idx in 0..x.nrows() {
+            let mut counter: Vec<u64> = vec![0; nclasses as usize];
+            for sp in &flattened {
+                // *counter_stats.entry(sp[sample_idx]).or_insert(0) += 1;
+                let single_pred = sp[sample_idx] as usize;
+                counter[single_pred] += 1;
+            }
+            let probas: Vec<f64> = counter.iter().map(|c| *c as f64 / ntrees as f64).collect();
+            result.push(Array1::from(probas));
+        }
+        Ok(result)
+    }
+}
+
+impl RandomForest {
+    pub fn fit(
+        hyperparameters: RandomForestParams,
+        x: &ArrayBase<impl Data<Elem = f64>, Ix2>,
+        y: &ArrayBase<impl Data<Elem = u64>, Ix1>,
+        max_n_rows: Option<usize>,
+    ) -> Self {
+        let n_estimators = hyperparameters.n_estimators;
+        let mut trees: Vec<DecisionTree> = Vec::with_capacity(n_estimators);
+        let single_tree_params = hyperparameters.tree_hyperparameters;
+        let max_n_rows = max_n_rows.unwrap_or_else(|| x.nrows());
+
+        //TODO check bootstrap
```

Comment:
Pretty important part of the algorithm is TODO here :smile: 
The bagging over samples looks good to me. When I was reading through the `sklearn` code I noticed a distinct absence of documentation internally. It could be nice to add a line or two here, something like,

Random forests work by bagging over samples (sampling from the original dataset with replacement) and taming a subset of features (sampling from the original features without replacement). Here we generate a random vector of sample indices `rnd_idx` and then selecting those rows from our original to get `xsample` and `ysample`.

Likewise here we'll need to generate something like a mask to decide whether to keep each feature. I'm actually rusty on how this is decided (i.e. whether it's just 50/50 whether to keep each feature or something else).

Reply:
Sounds like the norm is to uniformly select `sqrt(p)` features for classification or `floor(p/3)` for regression (without replacement) [reference](https://en.wikipedia.org/wiki/Random_forest#From_bagging_to_random_forests).

Reply:
With bagging, one just samples with replacement and re-run the single trees with full features. 
What you mention with `sqrt(p)` is the maximum number of features one can force each single tree to "focus" on.
The standard bagging implementation of random forest however, uses all available features. Which is why I left the  `_bootstrap` as TODO.

Reply:
cleaning up doc indeed it can be misleading

Reply:
If you don't take a subset of the features then it's not a random forest, it's just bagged decision trees.